### PR TITLE
[FIX] mail: no 'edited' at message post

### PR DESCRIPTION
**Pull request by alexkuhn**
Before this commit, when posting a new message in chatter that should be sent by emails (e.g. by mentioning the user that has notifications handled by email), then sometimes the message is considered as "(edited)" when it shouldn't.

This happens because when the new message must be sent by email, it creates a `mail.mail`, which inherits the newly created `mail.message`, and during the send process it writes on the `mail.message`, which changes the `write_date` and therefore it's different from `create_date`.

This commit fixes the issue by only changing `write_date` of `mail.message` when any write happens on some user-editable fields, i.e. comment and/or files and/or mentions.

### CODE REVIEW

File: addons/mail/models/mail_message.py [hunk 1]
```diff
@@ -690,7 +690,26 @@ def fetch(self, field_names):
...
+        self.flush_model()
+        for message in self:
+            super(Message, message)._write({ "write_date": write_date_by_id[message.id] })
+
+    def _write(self, vals):
+        prev_write_date_by_id = self._get_prev_write_date_by_id()
+        res = super(Message, self)._write(vals)
+        if not any(key in vals for key in ['body', 'attachment_ids', 'partner_ids']):
+            self._write_date(prev_write_date_by_id)
+        return res
```
Comment (rco-odoo) [inline addons/mail/models/mail_message.py:709]: This method is the one that generates the `UPDATE` queries in database.
Making it call `self._write_date()`, which will generate more updates is not smart at all.
Performance-wise, it's a disaster.

**Comment by tde-banana-odoo** [comment]
This seems a bit out of the blue ... and globally breaks the write_date semantic.

@rco-odoo note that the email sending (which notably write on notifications once it is sent) is done as a post-commit hook, I suppose this may have consequences. Could we suppress the write_date update just for that post-commit hook maybe ?

File: addons/mail/models/mail_message.py [hunk 2]
```diff
@@ -690,7 +690,26 @@ def fetch(self, field_names):
         self = self.sudo()
         return super().fetch(field_names)
 
+    def _get_prev_write_date_by_id(self):
+        prev_write_date_by_id = dict((message_id, {}) for message_id in self.ids)
+        for message in self:
+            prev_write_date_by_id[message.id] = message.write_date
+        return prev_write_date_by_id
```
Comment (rco-odoo) [inline addons/mail/models/mail_message.py:697]: Huh?  So much code for
```py
return {message.id: message.write_date for message in self}
```

**Review by rco-odoo — CHANGES_REQUESTED** [review]
Please come to our team with a precise description of the issue.
We should figure out a better way to do this.

**Comment by alexkuhn** [comment]
> @alexkuhn Could you check that it is the sending of emails or something else that triggers another write_date ? 

These triggers update on `write_date` of the `mail.message`:
- `mail.mail.py@create()` (fields `subject` and `mail_server_id`)
- `mail.mail.py@_send()` (fields `message_id`)
- `mail.mail.py@unlink()` (fields `write_uid` and `write_date` explicitly)

Everything happens in `mail_thread.py@_notify_thread_by_email`, as it creates and sends emails.

File: addons/mail/models/mail_message.py (hunk 1 above)
Comment (alexkuhn) [reply addons/mail/models/mail_message.py]: This is indeed poor performance-wise. My 1st imp was to conditionally not write on `write_date`, which was done with a key in context `skip_write_date`. This is ideal for performance, but without making any changes to `models.py` this is hardly feasible...

**Comment by seb-odoo** [comment]
I would hope instead of hacking write globally we could find one or some specific places in business code to set the write_date to what we need (either unchanged value, or plain reset to create date).

**Pull request closed without merging**